warehouse/templates/manage/token.html

```diff
@@ -0,0 +1,96 @@
+ # limitations under the License.
+-#}
+{% extends "manage_base.html" %}
+
+{% set user = request.user %}
+{% set title = "API token management" %}
+
+{% block account_active %}vertical-tabs__tab--is-active{% endblock %}
+{% block account_mobile_active %}vertical-tabs__tab--is-active{% endblock %}
+
+{% block title %}{{ title }}{% endblock %}
+
+{% block main %}
+  <h1 class="page-title">{{ title }}</h1>
+
+  {% if serialized_macaroon %}
+  <section id="provisioned-key">
+    <div class="callout-block callout-block--warning">
+      <p>
+        This token will only appear once. Copy it now.
+      </p>
+      <!-- TODO: This needs to be word-wrapped -->
+      <p>
+        {{ serialized_macaroon }}
+      </p>
+    </div>
+  </section>
+  {% endif %}
+
+  {% if user.macaroons|length > 0 %}
+  <section id="active-tokens">
+    <h2>Active API tokens</h2>
+    <table class="table table--light">
+      <thead>
+        <th>Description</th>
+        <th>Identifier</th>
+        <th>Created</th>
+        <th colspan="2">Last used</th>
+      </thead>
+      <tbody>
+        {% for macaroon in user.macaroons %}
+        <tr>
+          <td>{{ macaroon.description }}</td>
+          <td>{{ macaroon.id }}</td>
+          <td>{{ humanize(macaroon.created) }}</td>
+          <td>{{ humanize(macaroon.last_used) if macaroon.last_used else "Never" }}</td>
+          <td class="table__action">
+            <a href="#delete-token" class="button button--primary">Delete</a>
+            {% set title="Remove API token - " + macaroon.description %}
+            {% set confirm_button_label="Remove API token" %}
+            {% set action=request.route_path('manage.account.token.delete') %}
+            {% set extra_fields %}
+              <input type="text" name="macaroon_id" value="{{ macaroon.id }}" hidden>
+            {% endset %}
+            {{ confirm_modal(title=title, confirm_name="Username", confirm_string=user.username, confirm_button_label=confirm_button_label, slug="delete-token", extra_fields=extra_fields, action=action, warning=False, confirm_string_in_title=False) }}
+          </td>
+        </tr>
+        {% endfor %}
+      </tbody>
+    </table>
+  </section>
+  {% endif %}
+
+  <section id="add-token">
+    <h2>Create a new API token</h2>
+    <form method="POST" action="{{ request.route_path('manage.account.token.create') }}">
+      <input type="hidden" name="csrf_token" value="{{ request.session.get_csrf_token() }}">
+      <div class="form-group">
+        <input type="text" name="description">
+      </div>
+      <div class="form-group">
+        Scope:
+        <select name="token_scope">
+          <option value="scope:user">User (all projects)</option>
+        {% for project in project_names %}
+          <option value="scope:project:{{ project }}">{{ project }}</option>
+        {% endfor %}
+        </select>
+      </div>
+      <div>
+        <input value="Create token" class="button button--primary" type="submit">
```

Comment:
Right now, if I leave the token description empty, or put in a too-long (like, 40 characters) description, and click "Create token", I get taken back to the API token management page, but I don't get an error message. I figure we should probably give the user an error message.

Reply:
Yep, I need to expose the form errors. Will do that now, thanks!